File: 프로그래머스/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.py
def solution(elements):
    n = len(elements)
    # 원형 처리를 위해 배열 2번 이어 붙이기
    arr = elements * 2

    # 누적합(prefix sum) 배열 만들기
    prefix = [0] * (2 * n + 1)
    for i in range(2 * n):
        prefix[i + 1] = prefix[i] + arr[i]

    sums = set()

    # 부분 수열 길이 1 ~ n
    for length in range(1, n + 1):
        # 시작 인덱스 0 ~ n-1까지만 보면 원형의 모든 경우를 다 커버
        for start in range(n):
            total = prefix[start + length] - prefix[start]
            sums.add(total)

    return len(sums)

File: 프로그래머스/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.kt
class Solution {
    fun solution(elements: IntArray): Int {
        val n = elements.size

        // 원형 처리를 위해 배열을 2배로 늘림
        val arr = IntArray(2 * n) { i -> elements[i % n] }

        // prefix sum 배열
        val prefix = IntArray(2 * n + 1)
        for (i in 0 until 2 * n) {
            prefix[i + 1] = prefix[i] + arr[i]
        }

        val sums = HashSet<Int>()

        // 부분 수열 길이: 1 ~ n
        for (length in 1..n) {
            for (start in 0 until n) {
                val sum = prefix[start + length] - prefix[start]
                sums.add(sum)
            }
        }

        return sums.size
    }
}
